[AWS Lambda] permit tagging of lambda functions

fernandoalex opened the pull request:
##### SUMMARY
Add the ability to tag lambda functions

##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME
lambda.py

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
ansible 2.4.0 (detached HEAD f338dd2fc3) last updated 2017/05/18 14:26:07 (GMT +000)
  config file =
  configured module search path = [u'/home/ubuntu/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/ubuntu/ansible/lib/ansible
  executable location = /home/ubuntu/ansible/bin/ansible
  python version = 2.7.6 (default, Oct 26 2016, 20:30:19) [GCC 4.8.4]
```

fernandoalex committed:
permit tagging of lambda functions

Signed-off-by: fernandoalex <jfernandoalex@gmail.com>

fernandoalex committed:
changed iteritems for items and test for identity to is not

fernandoalex committed:
fixed task name in example

fernandoalex committed:
module object was not being passed to the function

Module object was not being passed to the funtions whitch was causing error in the error handling.

willthames commented on lib/ansible/modules/cloud/amazon/lambda.py:
```diff
@@ -227,6 +247,47 @@ def sha256sum(filename):
 
     return hex_digest
 
+def set_tag(client, module, tags, function):
```
There is a new compare_aws_tags function in module_utils.ec2 that should reduce the amount of work needed here

willthames commented on lib/ansible/modules/cloud/amazon/lambda.py:
```diff
@@ -410,6 +473,11 @@ def main():
 
+        # Tag Function
+        if tags is not None and len(tags) >= 0:
+            if set_tag(client, module, tags, current_function):
```
`tag_resource` is *very* new to botocore, we cannot assume everyone has it.

That might not even be a ClientError, it might be an AttributeError.

Can you test with botocore 1.5.39 and see how this behaves. It's fine to fail if someone is passing tags with an old version of botocore, but we should say "Could not set tags - please upgrade to botocore 1.5.40 or higher"

```
    if tags is not None:
        if not hasattr(client, 'tag_resource'):
            module.fail_json(msg="Tags are not supported by your version of botocore. "
                             "Please update to 1.5.40 or newer")
```

willthames reviewed (changes requested):
Need to handle older botocore more clearly

willthames commented on lib/ansible/modules/cloud/amazon/lambda.py:
```diff
@@ -113,6 +113,12 @@
     version_added: "2.3"
+  tags:
+    description:
+      - tag dict to apply to the function
```
Requires botocore 1.5.40 or above

willthames commented on lib/ansible/modules/cloud/amazon/lambda.py:
```diff
@@ -486,6 +554,11 @@ def main():
+            if set_tag(client, module, tags, get_current_function(client, name)):
+                changed = True
+
         response = get_current_function(client, name, qualifier=current_version)
```
get_current_function should be updated to return tags in ansible form

s-hertel committed:
Use compare_aws_tags

Check for attribute and fail if botocore needs to be updated

Update version_added

s-hertel committed:
Fix examples

s-hertel committed:
pep8

s-hertel committed:
Remove unnecessary var

Handle BotoCoreError
